### PR TITLE
chore(netlify-cms): update for v2, add standalone build system

### LINKED ISSUE #4516: Supporting plugins that extract stylesheets for a separate entry point
## Description

Plugins that modify Webpack CSS and CSS preprocessor file loading may overwrite styles output by Gatsby's default CSS and CSS module loaders for production builds.

## Why this is happening

The Netlify CMS plugin is the first official plugin to utilize `extract-text-webpack-plugin` for it's own entry point (#3611). Prior to this, Gatsby used a **single** instance of the extract plugin, but supporting plugins that output a separate stylesheet requires using **multiple** instances (#3652). The plugins that load styles for the main style bundle, e.g. sass, stylus, etc., rely on the extract plugin added by Gatsby rather than adding their own, but this stops working correctly when there are multiple extract plugins, an issue that surfaced for projects using both the Netlify CMS plugin and any other plugin that implemented the webpack extract plugin (#4335).

## How to address

In #4495, a solution is reached where the Sass plugin retrieves and utilizes the first instance of `extract-text-webpack-plugin` in the plugins array. This ensures that general site styles that are meant for the main bundle get processed by Gatsby's own extract plugin instance, regardless of any other instances that plugins may have added. This approach relies on plugins never intentionally splicing in their own extract plugin instance _before_ Gatsby's instance. An approach that does not rely on this will involve a fair amount of additional code in all styling plugins.

This solution will need to be implemented in all plugins that utilize `extract-text-webpack-plugin`, or else these plugins will not work with plugins that use their own extract plugin instance.

## Options

I see two ways to do this:

1. Somehow make it easier to get and reuse Gatsby's current stage extract plugin instance for simplicity, and to avoid guessing
2. Implement the fix from #4495 

Whatever approach we choose, it will need to be applied to all styling plugins that utilize the extract plugin. Note that none of this points to a breaking change, but a change that must be made if these plugins are to be used in conjunction with other plugins that utilize the extract plugin for separate entry points.

### LINKED ISSUE #6048: [v2] Using gatsby-plugin-netlify-cms in v2 doesn't work correctly
## Description
I am using gatsby@next and I am trying to get the netlify-cms plugin to work.
Using the 1.0 plugin doesn't get inserted into webpack and the index.html in admin doesn't get build.
Using the 2.x plugin causes a error when executing the onCreateWebpackConfig
The code looks like this
```
exports.onCreateWebpackConfig = ({ config, stage }, { modulePath }) => {
  config.merge({
    entry: {
      cms: [`${__dirname}/cms.js`, modulePath].filter(p => p),
    },
    plugins: plugins(stage),
  })

  module(config, stage)

  return config
}
```
And this seems to be an older pre final 2 api. The new API uses actions and not config, state and modulePath.

Describe the issue that you're seeing.

The error is:
```
error UNHANDLED EXCEPTION


  TypeError: Cannot read property 'merge' of undefined
  
  - gatsby-node.js:91 Object.exports.onCreateWebpackConfig
    [gatsby-starter-netlify-cms]/[gatsby-plugin-netlify-cms]/gatsby-node.js:91:10
  
  - api-runner-node.js:104 runAPI
    [gatsby-starter-netlify-cms]/[gatsby]/dist/utils/api-runner-node.js:104:37
  
  - api-runner-node.js:173 mapSeries
    [gatsby-starter-netlify-cms]/[gatsby]/dist/utils/api-runner-node.js:173:25
```

### Steps to reproduce
Configure a 2.x gatsby site with the 2.x version of the gatsby-plugin-netlify-cms plugin.

### Expected result
It would generate the appropriate index.html file in admin.

### Actual result
It just errs out during the build

### Environment
MacOS, node v9.11.1, npm v6.1.0, yarn v1.7.0

  System:
    OS: macOS High Sierra 10.13.5
    CPU: x64 Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 9.11.1 - ~/.nvm/versions/node/v9.11.1/bin/node
    Yarn: 1.7.0 - /usr/local/bin/yarn
    npm: 6.1.0 - ~/.nvm/versions/node/v9.11.1/bin/npm
  Browsers:
    Chrome: 67.0.3396.87
    Firefox: 60.0.1
    Safari: 11.1.1
  npmPackages:
    gatsby: next => 2.0.0-beta.2 
    gatsby-image: next => 2.0.0-beta.1 
    gatsby-plugin-emotion: next => 2.0.0-beta.1 
    gatsby-plugin-manifest: next => 2.0.2-beta.1 
    gatsby-plugin-netlify: next => 2.0.0-beta.1 
    gatsby-plugin-netlify-cms: next => 2.0.0-beta.1 
    gatsby-plugin-offline: next => 2.0.0-beta.1 
    gatsby-plugin-react-helmet: next => 3.0.0-beta.1 
    gatsby-plugin-sharp: next => 2.0.0-beta.1 
    gatsby-source-filesystem: next => 2.0.1-beta.1 
    gatsby-transformer-remark: next => 2.1.1-beta.1 
    gatsby-transformer-sharp: next => 2.1.1-beta.1 
    gatsby-transformer-yaml: next => 2.1.1-beta.1 
  npmGlobalPackages:
    gatsby: 2.0.0-beta.1

### File contents (if changed)

`gatsby-config.js`: N/A 
`package.json`: N/A 
`gatsby-node.js`: N/A 
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

### PR DESCRIPTION
**Pull request by erquhart**
Fixes #4516, #6048, #6248.

* Update to v2 API's
* Use a standalone webpack compiler to ensure against impacts on main site
* Accept custom preview pane styles (even raw preprocessor styles) via plugin options
* Optionally disable identity widget (ported from #5942)
* Make custom preview templates and styles work during local development
* Added a few other small options (documented)

### CODE REVIEW

**Comment by erquhart** [comment]
Not sure why those failures are happening, not seeing them locally.

**Comment by erquhart** [comment]
That's fair, I'll switch it to webpack.

**Comment by emanueleperuffo** [comment]
I've created PR #6391

**Comment by erquhart** [comment]
@emanueleperuffo I mentioned this on your PR, but THANK YOU so much for digging into this. Our efforts just happened to overlap as the Gatsby/Netlify CMS use case has been one of the "fires I've let burn" for a bit while tending to other things.

I took the opportunity to finish the v2 update, and also to fix longstanding issues with using custom preview pane styles during Gatsby local development.

There are a lot of known issues with this plugin, especially as documented in #4516, and I _think_ this PR addresses all known issues while keeping the plugin minimally complex.

@m-allanson all finished here.

**Comment by erquhart** [comment]
The test and deploy failures are present in other pull requests too, I don't believe any of them are related to this PR.

File: packages/gatsby-plugin-netlify-cms/src/cms.js [hunk 1]
```diff
@@ -1,8 +1,6 @@
-/* eslint-disable no-unused-vars */
-/* eslint-env browser */
 import CMS from "netlify-cms"
 import "netlify-cms/dist/cms.css"
-import netlifyIdentityWidget from "netlify-identity-widget"
 
-window.netlifyIdentity = netlifyIdentityWidget
-netlifyIdentityWidget.init()
+if (NETLIFY_CMS_PREVIEW_STYLES_SET) {
```
Comment (ascorbic) [inline packages/gatsby-plugin-netlify-cms/src/cms.js]: The CI builds are failing because this is an undefined variable.

Comment (erquhart) [reply packages/gatsby-plugin-netlify-cms/src/cms.js]: Ah, I see that now. Weird that it passes linting locally.

**Comment by erquhart** [comment]
@m-allanson @KyleAMathews I don't believe this PR is responsible for any test failures - can this be merged?

**Comment by KyleAMathews** [comment]
Hmm can't recreate the test errors locally... @m-allanson or @pieh — any idea what those test errors mean?

**Comment by m-allanson** [comment]
It turns out that React isn't included as a dependency anywhere in the repo. The `netlify-cms` package depends on React, but that is now a peerDependency. I guess this wasn't showing up locally as the old React packages would still be hanging out in `node_modules`

@erquhart I got this working by adding `react` and `react-dom` as devDepencies of `packages/gatsby`. I have a commit locally but don't have access to push it to your repo. Feel free to create your own commit for this instead.

**Comment by m-allanson** [comment]
> this PR addresses all known issues while keeping the plugin minimally complex.

Just wanted to add a big 🎉 for this :)

What is the relation between this PR and #6391? Should just this one be merged?

**Comment by shaunbent** [comment]
Looking forward to seeing this PR merged, this is currently blocking us from upgrading our site to Gatsby V2.

Thanks for the awesome work @erquhart 🎉😎

**Comment by KyleAMathews** [comment]
Ahhhh! That makes sense @m-allanson! Weird though you can't push to his repo as I can :thinking:

Anyways, I'll make the change and push it right now.

**Comment by KyleAMathews** [comment]
Published PR to gatsby-plugin-netlify-cms@2.0.0-beta.6

**Comment by steverandy** [comment]
Previously we had this line
```
netlifyIdentityWidget.on(`login`, () => {
  document.location.href = `/admin/`
})
```
Is there any reason why now it's changed to reloading the page?

**Comment by ascorbic** [comment]
It seems that this is no longer transpiling TypeScript. My preview templates are TypeScript, and previously this was fine because it used the webpack config from the main site, which includes the TypeScript plugin. Now they're failing to compile. Is there a way of exposing the Netlify Webpack config?

**Comment by erquhart** [comment]
@steverandy I was actually in the middle of troubleshooting why reloads weren't happening on login, that needs to be changed.

@ascorbic the Netlify CMS webpack config uses any plugins that loaded _before_ the Netlify CMS plugin itself. Try moving your Typescript plugin before the Netlify CMS plugin and see if that works. It would be worth considering how to capture all plugins used, regardless of placement, but the current setup should work for most.

**Comment by emanueleperuffo** [comment]
Is it possible to use version > 2.0.1 like 2.0.1-beta.6 so npm doesn't update the package to 2.0.1 which is for gatsby v1?

**Comment by pieh** [comment]
@emanueleperuffo Hmmm, we should bump major version for v2 - we did that some time ago for packages in this repository, but it seems like we have missed this package

**Comment by ascorbic** [comment]
Is anyone able to get this to generate the `styles.css` for custom preview styles? I opened a separate issue, #6641, as I thought it may have been a documentation issue, but I'm starting to think it may be a problem with the webpack config.

**Pull request merged**